README.md
***************************************************************************
**    multiblend
***************************************************************************

multiblend (c) 2014 David Horman

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA.

The author can be contacted at: david.horman@jerseymail.co.uk

***************************************************************************
**    libtiff
***************************************************************************

Copyright (c) 1988-1997 Sam Leffler
Copyright (c) 1991-1997 Silicon Graphics, Inc.

Permission to use, copy, modify, distribute, and sell this software and 
its documentation for any purpose is hereby granted without fee, provided
that (i) the above copyright notices and this permission notice appear in
all copies of the software and related documentation, and (ii) the names of
Sam Leffler and Silicon Graphics may not be used in any advertising or
publicity relating to the software without the specific, prior written
permission of Sam Leffler and Silicon Graphics.

THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  

IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
OF THIS SOFTWARE.

***************************************************************************
**    libjpeg-turbo
***************************************************************************

libjpeg-turbo is licensed under a non-restrictive, BSD-style license
(see README.)  The TurboJPEG/OSS wrapper (both C and Java versions) and
associated test programs bear a similar license, which is reproduced below:

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

- Redistributions of source code must retain the above copyright notice,
  this list of conditions and the following disclaimer.
- Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.
- Neither the name of the libjpeg-turbo Project nor the names of its
  contributors may be used to endorse or promote products derived from this
  software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS",
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.



changelog.txt
v0.6.2
* fix to ignore some enblend options possibly used by default by Hugin
* fix for one minor compiler warning

v0.6.1
* fix for a bug that might cause black banding on the bottom edge

v0.6
* option to cache input images (--cache)
* fix for vignetting and corruption of right edge
* fixes to 16-bit processing (spotted by Matthias Kabel)
* nicer colouring for --save-seams

v0.5
* fixes to GeoTIFF support
* fix for dark banding problems when compiling with g++>=4.7

v0.4
* --save-seams and --load-seams for external editing of seams
* TIFF position offsets are now preserved
* removal of some image size limits for 64 bit builds (based on changes by Pablo dAngelo)
* (untested) GeoTIFF support (based on changes by Pablo dAngelo)

v0.31beta
* pseudowrapping bugs fixed

v0.3beta
* pseudowrapping - single-image input is blended around the left/right image boundary

v0.2
* fixed: no more segfaults when compiling with -O3
* work-around for seaming bug which left areas smeared in some cases
* added --reverse to change image priority (affects how "seaming bug" areas are solved)
* added JPEG output support

v0.1a
* added --bgr switch to correct channel order for some users

v0.1
* initial release



multiblend.cpp
/*
  multiblend (c) 2013 David Horman

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2, or (at your option)
  any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA.

  The author can be contacted at: david.horman@jerseymail.co.uk

  Discussion at http://tawbaware.com/forum2/viewtopic.php?f=3&t=6396
             or http://groups.google.com/group/hugin-ptx/
*/

#include <algorithm>
using namespace std;
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef __APPLE__
  #define memalign(a,b) malloc((b))
#else
  #include <malloc.h>
#endif
#include <emmintrin.h>

#ifdef WIN32
#define NOMINMAX
#include <Windows.h>
#endif

#define PY(i,l) g_images[i].pyramid[l]

#include <stdarg.h>
#include <jpeglib.h>
#include <png.h>
#include <tiffio.h>
#include "globals.cpp"
#include "functions.cpp"
#include "geotiff.cpp"
#include "loadimages.cpp"
#include "seaming.cpp"
#include "maskpyramids.cpp"
#include "blending.cpp"
#include "write.cpp"
#include "pseudowrap.cpp"
#include "go.cpp"

#ifdef WIN32
#pragma comment(lib,"libtiff.lib")
#pragma comment(lib,"turbojpeg-static.lib")
#pragma comment(lib,"libpng.lib")
#pragma comment(lib,"zlib.lib")
#endif

void help() {
	printf("Usage: multiblend [options] [-o OUTPUT] INPUT...\n");
	printf("Options:\n");
	printf("   -l X                  X > 0: limit number of blending levels to x\n");
	printf("                         X < 0: reduce number of blending levels by -x\n");
	printf("   -d DEPTH              override automatic output image depth (8 or 16)\n");
	printf("  --nocrop               do not crop output\n");
	printf("  --bgr                  swap RGB order\n");
	printf("  --wideblend            calculate number of levels based on output image size,\n");
	printf("                         rather than input image size\n");
	printf("  --compression=X        output file compression. For TIFF output, X may be:\n");
	printf("                         NONE (default), PACKBITS, or LZW\n");
	printf("                         For JPEG output, X is JPEG quality (0-100, default 75)\n");
	printf("  --cache                cache input images to disk to minimise memory usage\n");
  printf("  --save-seams <file>    Save seams to PNG file for external editing\n");
  printf("  --no-output            Don't perform blend (for use with --save-seams)\n");
  printf("  --load-seams <file>    Load seams from PNG file\n");
	printf("  --bigtiff              BigTIFF output (not well tested)\n");
	printf("  --reverse              reverse image priority (last=highest)\n");
	printf("  --quiet                suppress output (except warnings)\n");
  printf("\n");
  printf("Pass a single image as input to blend around the left/right boundary.\n");
	exit(0);
}

int main(int argc, char* argv[]) {
  int i;
  int input_args;
	int temp;
	my_timer timer_all;

	timer_all.set();
	TIFFSetWarningHandler(NULL);

	printf("\n");
	printf("multiblend v0.6.2 (c) 2014 David Horman        http://horman.net/multiblend/\n");
	printf("----------------------------------------------------------------------------\n");

  if (argc==1 || !strcmp(argv[1],"-h") || !strcmp(argv[1],"--help") || !strcmp(argv[1],"/?")) help();

  printf("\n");

  if (argc<3) die("not enough arguments (try -h for help)");

	for (i=1; i<argc-1; i++) {
    if (!strcmp(argv[i],"-d")) {
			g_workbpp_cmd=atoi(argv[++i]);
			if (g_workbpp_cmd!=8 && g_workbpp_cmd!=16) {
				die("invalid output depth specified");
			}
		}
		else if (!strcmp(argv[i],"-l")) {
			temp=atoi(argv[++i]);
			if (temp>=0) g_max_levels=max(1,temp); else g_sub_levels=-temp;
		}
		else if (!strcmp(argv[i],"--nomask")) g_nomask=true;
		else if (!strcmp(argv[i],"--nocrop")) g_crop=false;
		else if (!strcmp(argv[i],"--bigtiff")) g_bigtiff=true;
		else if (!strcmp(argv[i],"--bgr")) g_bgr=true;
		else if (!strcmp(argv[i],"--wideblend")) g_wideblend=true;
		else if (!strcmp(argv[i],"--noswap")) g_swap=false;
		else if (!strcmp(argv[i],"--reverse")) g_reverse=true;
		else if (!strcmp(argv[i],"--timing")) g_timing=true;
		else if (!strcmp(argv[i],"--dewhorl")) g_dewhorl=true;

		else if (!strcmp(argv[i],"-w")) output(0,"ignoring enblend option -w\n");
		else if (!strncmp(argv[i],"-f",2)) output(0,"ignoring enblend option -f\n");
		else if (!strcmp(argv[i],"-a")) output(0,"ignoring enblend option -a\n");
		else if (!strcmp(argv[i],"--no-ciecam")) output(0,"ignoring enblend option --no-ciecam\n");
		else if (!strncmp(argv[i],"--primary-seam-generator",24)) output(0,"ignoring enblend option --primary-seam-generator\n");

		else if (!strncmp(argv[i],"--compression",13)) {
			char* comp=argv[i]+14;
			if (strcmp(comp,"0")==0) g_jpegquality=0;
			else if (atoi(comp)>0) g_jpegquality=atoi(comp);
			else if (_stricmp(comp,"lzw")==0) g_compression=COMPRESSION_LZW;
			else if (_stricmp(comp,"packbits")==0) g_compression=COMPRESSION_PACKBITS;
//			else if (_stricmp(comp,"deflate")==0) g_compression=COMPRESSION_DEFLATE;
			else if (_stricmp(comp,"none")==0) g_compression=COMPRESSION_NONE;
			else die("unknown compression codec");
		}
    else if (!strcmp(argv[i],"-v") || !strcmp(argv[i],"--verbose")) g_verbosity++;
		else if (!strcmp(argv[i],"-q") || !strcmp(argv[i],"--quiet")) g_verbosity--;
    else if (!strcmp(argv[i],"--debug")) g_debug=true;
		else if (!strcmp(argv[i],"--saveseams") || !strcmp(argv[i],"--save-seams")) g_seamsave_filename=argv[++i];
    else if (!strcmp(argv[i],"--loadseams") || !strcmp(argv[i],"--load-seams")) g_seamload_filename=argv[++i];
		else if (!strcmp(argv[i],"--simpleseam") || !strcmp(argv[i],"--simple-seam")) g_simpleseam=true;
    else if (!strcmp(argv[i],"--savemasks") || !strcmp(argv[i],"--save-masks")) g_savemasks=true;
    else if (!strcmp(argv[i],"--saveoutpyramids")) g_save_out_pyramids=true;
		else if (!strcmp(argv[i],"--savexor") || !strcmp(argv[i],"--save-xor")) g_xor_filename=argv[++i];
    else if (!strcmp(argv[i],"--no-output")) g_nooutput=true;
		else if (!strcmp(argv[i],"--cache")) g_caching=true;
    else if (!strcmp(argv[i],"-o") || !strcmp(argv[i],"--output")) {
      g_output_filename=argv[++i];
			char* ext=strrchr(g_output_filename,'.')+1;
			if (!(_stricmp(ext,"jpg") && _stricmp(ext,"jpeg"))) {
				if (g_compression!=-1) {
					output(0,"warning: JPEG output; ignoring TIFF compression setting\n");
					g_compression=-1;
				}
				if (g_jpegquality==-1) g_jpegquality=75;
			} else if (!(_stricmp(ext,"tif") && _stricmp(ext,"tiff"))) {
				if (g_jpegquality!=-1) {
					output(0,"warning: TIFF output; ignoring JPEG quality setting\n");
					g_jpegquality=-1;
				} else if (g_compression==-1) {
					g_compression=COMPRESSION_NONE;
				}
			} else {
				die("unknown file extension!");
			}

      i++;
      break;
    } else {
      die ("unknown argument \"%s\"",argv[i]);
    }
  }

  if (!g_output_filename && !g_seamsave_filename) die("no output file specified");

  if (!strcmp(argv[i],"--")) i++;

  input_args=argc-i;
  if (input_args==0) die("no input files specified");
  if (input_args>255) die("too many (>255) input images specified");

	go(&argv[i],input_args);

  if (g_timing) timer_all.report("Execution time");

	clear_temp();

  if (g_debug) {
		printf("\nPress Enter to end\n");
		getchar();
	}

	exit(0);
	return 0;
}
